contracts/BarterMarket.sol

```diff
@@ -148,15 +148,113 @@ contract BarterWalletFactory {
+		// If anything fails all state changes will be reverted anyways, kind of like a db transaction
+		offers[_offerId].state = State.ACCEPTED;
+
+		// Check that the target's assets are approved for the trade, then transfer
+		CoinBundle memory askCoins = offer.askBundle.tokens;			
+		NFTBundle memory askNfts = offer.askBundle.nfts;
+
+		for (uint256 i = 0; i < askCoins.contractAddresses.length; i++) {
+			ERC20 coinContract = ERC20(askCoins.contractAddresses[i]);
+			
+			require(
+				coinContract.balanceOf(offer.target) >= askCoins.amounts[i],
+				"Not enough tokens"
+			);
+			require(
+				coinContract.allowance(offer.target, address(this)) >=
+					askCoins.amounts[i] +
+						committedTokens[offer.target][askCoins.contractAddresses[i]],
+				"Not enough allowed tokens"
+			);
+
+			coinContract.transferFrom(offer.target, offer.offerer, askCoins.amounts[i]);
+		}
+
+		for (uint256 i = 0; i < askNfts.contractAddresses.length; i++) {
+			ERC721 nftContract = ERC721(askNfts.contractAddresses[i]);
+
+			require(
+				nftContract.ownerOf(askNfts.ids[i]) == msg.sender,
+				"You no longer own this nft"
+			);
+			require(
+				nftContract.isApprovedForAll(msg.sender, address(this)),
+				"Not approved for all NFT transfers"
+			);
+			require(
+				committedNFTs[askNfts.contractAddresses[i]][askNfts.ids[i]] == false,
+				"Already committed to another offer"
+			);
+
+			nftContract.safeTransferFrom(offer.target, offer.offerer, askNfts.ids[i]);
+		}
+
+		// Double check offerer's assets are approved as approvals can be revoked, then transfer
+		CoinBundle memory offerCoins = offer.offerBundle.tokens;
+		NFTBundle memory offerNfts = offer.offerBundle.nfts;
+
+		for (uint256 i = 0; i < offerCoins.contractAddresses.length; i++) {
+			ERC20 coinContract = ERC20(offerCoins.contractAddresses[i]);
+
+			require(
+				coinContract.balanceOf(offer.offerer) >= offerCoins.amounts[i],
+				"Not enough tokens"
+			);
+			require(
+				coinContract.allowance(offer.offerer, address(this)) >=
+					offerCoins.amounts[i],
+				"not enough allowed tokens"
+			);
+
+			committedTokens[offer.offerer][offerCoins.contractAddresses[i]] -= offerCoins.amounts[i];
+			coinContract.transferFrom(offer.offerer, offer.target, offerCoins.amounts[i]);
+		}
+
+    for (uint256 i = 0; i < offerNfts.contractAddresses.length; i++) {
+      ERC721 nftContract = ERC721(offerNfts.contractAddresses[i]);
+
+      require(
+        nftContract.ownerOf(offerNfts.ids[i]) == offer.offerer,
+        "Not the offerers NFT anymore"
+      );
+      require(
+        nftContract.isApprovedForAll(offer.offerer, address(this)),
+        "Not approved for all NFT transfers"
+      );
+
+			committedNFTs[offerNfts.contractAddresses[i]][offerNfts.ids[i]] = false;
+			nftContract.transferFrom(offer.offerer, offer.target, offerNfts.ids[i]);
+    }
+
+		require(
+       offer.offerBundle.offeredEther <= address(this).balance,
+      "Not enough eth in contract!"
+    );
+
```

Comment:
I wonder if we should do a check if the assets were transferred here.  Could someone could write a malicious ERC20 or ERC721 with a transfer function that doesn't actually change ownership?  Is that something we need to think about?